### PR TITLE
UCT/IB: Enable pci relaxed order on Fujitsu ARM by default

### PR DESCRIPTION
**Pull request by Artemy-Mellanox**

### CODE REVIEW

**Comment by mellanox-github** [comment]
<details><summary>Mellanox CI: <b>PASSED</b#  > on 13 workers (click for details)</summary>

**Note:** the logs will be deleted after 07-Apr-2020

Agent/Stage | Status
-- | --
[_main](https://gist.githubusercontent.com/mellanox-github/f565e34e628b85234e8fefa8b70cc3f2/raw/5f233f8dd25a69fe8a520af794f30745ebaf6b8a/_main) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/f565e34e628b85234e8fefa8b70cc3f2/raw/43e26e828b4bdf80f83b3c6a7e3cfc04b54e22f5/hpc-arm-cavium-jenkins_W0) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/f565e34e628b85234e8fefa8b70cc3f2/raw/c4e992b24188d3c46f5c8faafcacb9d7a45ba5d3/hpc-arm-cavium-jenkins_W1) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/f565e34e628b85234e8fefa8b70cc3f2/raw/15204b3a46fc3939c94ce1325c7f8941b74e77a8/hpc-arm-cavium-jenkins_W2) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/f565e34e628b85234e8fefa8b70cc3f2/raw/1978a8b54484327ea2c70e3a34d755f273c456b0/hpc-arm-cavium-jenkins_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W0](https://gist.githubusercontent.com/mellanox-github/f565e34e628b85234e8fefa8b70cc3f2/raw/054e8d91169e4eda40f02cc95e2f612b9c9cdd8d/hpc-test-node-new_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W1](https://gist.githubusercontent.com/mellanox-github/f565e34e628b85234e8fefa8b70cc3f2/raw/6ffe514d07009c86e538a80be1af8b87999e2aca/hpc-test-node-new_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W2](https://gist.githubusercontent.com/mellanox-github/f565e34e628b85234e8fefa8b70cc3f2/raw/624f717277ffb03a44c638450002bea932120a75/hpc-test-node-new_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W3](https://gist.githubusercontent.com/mellanox-github/f565e34e628b85234e8fefa8b70cc3f2/raw/8a1831f7d10fbd48b8e2ad9b59fff64c9f7bf9db/hpc-test-node-new_W3) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/f565e34e628b85234e8fefa8b70cc3f2/raw/91f71cb32b332ade4d33a123b8d11cebcecf6f60/r-vmb-ppc-jenkins_W0) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/f565e34e628b85234e8fefa8b70cc3f2/raw/5856f68d20a646b4f602ced29bba587a3b95cdc5/r-vmb-ppc-jenkins_W1) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/f565e34e628b85234e8fefa8b70cc3f2/raw/e240ea98a5a974a12d5ececb188392709eff9e1e/r-vmb-ppc-jenkins_W2) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/f565e34e628b85234e8fefa8b70cc3f2/raw/7c62d4115355468218777c2990709f005230978c/r-vmb-ppc-jenkins_W3) | :heavy_check_mark: SUCCESS

File: src/uct/ib/base/ib_md.c [hunk 1]
```diff
@@ -1482,6 +1482,23 @@ ucs_status_t uct_ib_md_open_common(uct_ib_md_t *md,
...
     md->super.component = &uct_ib_component;
 
+    if ((md->config.mr_relaxed_order == UCS_CONFIG_ON) && !IBV_ACCESS_RELAXED_ORDERING) {
+        ucs_warn("relaxed order memory access requested but not supported");
+    } else if (md->config.mr_relaxed_order == UCS_CONFIG_AUTO) {
+#ifdef __aarch64__
+        ucs_aarch64_cpuid_t cpuid;
+        ucs_aarch64_cpuid(&cpuid);
+
+        if ((cpuid.implementer == 0x46) && (cpuid.architecture == 8)) {
```
Comment (shamisp) [inline src/uct/ib/base/ib_md.c:1492]: Why you wound not add some sort of generic query about PCIE ordering preferences in ucs/arch ?

Comment (yosefe) [reply src/uct/ib/base/ib_md.c:1492]: in UCS?

File: src/ucs/arch/aarch64/cpu.h [hunk 2]
```diff
@@ -119,6 +119,12 @@ static inline ucs_cpu_model_t ucs_arch_get_cpu_model()
 
 static inline ucs_cpu_vendor_t ucs_arch_get_cpu_vendor()
 {
+    ucs_aarch64_cpuid_t cpuid;
+    ucs_aarch64_cpuid(&cpuid);
+
+    if ((cpuid.implementer == 0x46) && (cpuid.architecture == 8))
+        return UCS_CPU_VENDOR_FUJITSU_ARM;
```
Comment (yosefe) [inline src/ucs/arch/aarch64/cpu.h:126]: { }

File: src/ucs/arch/cpu.h [hunk 3]
```diff
@@ -143,6 +144,21 @@ static inline void ucs_clear_cache(void *start, void *end)
...
+{
+    if (ucs_arch_get_cpu_vendor() == UCS_CPU_VENDOR_FUJITSU_ARM) {
+        return 1;
+    }
+
+    if (ucs_arch_get_cpu_model() == UCS_CPU_MODEL_AMD_ROME) {
+        return 1;
+    }
+
+    return 0;
```
Comment (yosefe) [inline src/ucs/arch/cpu.h:158]: return model() == FUJITSU_ARM
for AMD ROME we need to test it first

File: src/uct/ib/base/ib_md.c [hunk 4]
```diff
@@ -1485,7 +1485,11 @@ ucs_status_t uct_ib_md_open_common(uct_ib_md_t *md,
     if ((md->config.mr_relaxed_order == UCS_CONFIG_ON) && !IBV_ACCESS_RELAXED_ORDERING) {
         ucs_warn("relaxed order memory access requested but not supported");
     } else if (md->config.mr_relaxed_order == UCS_CONFIG_AUTO) {
-        md->config.mr_relaxed_order = UCS_CONFIG_OFF;
+        if (ucs_cpu_prefer_relaxed_order()) {
+            md->config.mr_relaxed_order = UCS_CONFIG_ON;
```
Comment (yosefe) [inline src/uct/ib/base/ib_md.c:1489]: why are we sometimes using md_config and other cases md->config?

Comment (Artemy-Mellanox) [reply src/uct/ib/base/ib_md.c:1489]: AFAIU md->config is part of md_config which will be accessible after md open. Isn't it https://github.com/openucx/ucx/commit/21ac89b7008a0e9400d0afcb243ca678c956ef4e?

Comment (yosefe) [reply src/uct/ib/base/ib_md.c:1489]: so, we should move mr_relaxed_order from uct_ib_md_ext_config_t to uct_ib_md_config_t, then add "int relaxed_order" to uct_md_t - the difference is it's not tri-state, but bool

**Pull request merged**